External http01 solver for internal services

kinolaev opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**: to get letsencrypt's certificates for internal services.

Situation:

- two ingress controllers with different ingress classes, one for internal services (no access from internet) and one for external.
- two different dns servers, internal and external.
- dns01 challenge is not possible
- certificate from letsencrypt needed for internal service

Proposed solution: allow create http01 challenge solver ingress resource with external ingress class for ingress resources with internal ingress classes.

**Example**

- internal ingress controller with loadBalancerIP 10.10.10.10  
- external ingress controller with loadBalancerIP 10.10.10.11  
- `myservice.company.org` on internal dns servers resolves to 10.10.10.10  
- `myservice.company.org` on external dns servers resolves to 1.1.1.1 (nat 10.10.10.11)  
- ingress resource for `myservice.company.org` with `kubernetes.io/ingress.class: internal` annotation
- LetEncrypt staging issuer resource:
```yaml
apiVersion: certmanager.k8s.io/v1alpha1
kind: ClusterIssuer
metadata:
  name: letsencrypt-staging
spec:
  acme:
    email: email@example.com
    server: https://acme-staging-v02.api.letsencrypt.org/directory
    http01:
      ingressClassRules:
      - matchValues:
        - internal
        - ""
        ingressClass: external
      - ingressClass: external
        ingressCheck: true
  privateKeySecretRef:
      name: letsencrypt-staging
```

`ingressCheck` allows to send check requests to `.status.loadbalancer.ingress` of solver ingress resource instead of `myservice.company.org` (because `myservice.company.org` on internal dns servers resolves to internal ingress controller but solver is accessible only through external).

`matchValues` specifies ingress classes for which solver ingress class should be set to `ingressClass` field value. `""` - matches if ingress resource has no annotation. Empty `matchValues` means all ingress classes.

Result: when adding tls config and `certmanager.k8s.io/cluster-issuer: letsencrypt-staging` annotation to ingress resource for `myservice.company.org` ingress resource with `kubernetes.io/ingress.class: external` annotation will be created to solve http01 challenge.

**Special notes for your reviewer**: sorry for my english

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

**Updated 11 oktober 2018**: replace `skipCheck` with `ingressCheck`

kinolaev commented:
/assign @kragniz

munnerz commented:
Thanks for the detailed explanation of what you're trying to do here!

So as it stands, I think these API changes are not required for you to achieve this setup. You should be able to manually create a Certificate like so:

```yaml
apiVersion: certmanager.k8s.io/v1alpha1
kind: Certificate
metadata:
  name: my-cert
spec:
  secretName: my-cert-tls
  dnsNames:
  - myservice.company.org
  issuerRef:
    name: letsencrypt-staging
    kind: ClusterIssuer
  acme:
    config:
    - domains:
      - myservice.company.org
      http01:
        ingressClass: external
```

cert-manager will then create ingress resources with the 'external' class set in order to solve the challenge.

If you then configure cert-manager to use an *external* DNS resolver, instead of your internal resolver, it should perform a DNS lookup for `myservice.company.org` which will resolve to 1.1.1.1. You can do this using the 'pod dns config' options, which we expose via the Helm chart: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods-dns-policy

I think the key difference here, and what your PR implements, is supporting this with *ingress-shim*. At the moment, I don't think we'd want to further complicate the Issuer resource type in order to support this use case when there are alternative ways to achieve what you're after (i.e. manually specifying the Certificate).

Does that make sense?

kinolaev commented:
Hello @munnerz!
I updated PR, ingressClassRules now stored as json in ConfigMap `clusterResourceNamespace/ingress-shim`, key `ingress-class-rules`.

kinolaev commented:
Hello @munnerz!

Can you comment on updated PR and answer questions above? Or you don't completly interesting in this feature and I need to find another people/companies to ensure that my case is not unique?

kinolaev commented:
Hello @munnerz!

We partly fixed our problem by creating proxy server that have access to the internet and use external DNS servers.

To allow ingress-shim to use certificate ingress class other then original ingress class I created PR #1006.

munnerz commented:
@kinolaev great 😄

Looking forward to getting that support in!

kinolaev commented:
I updated PR:
- `IngressCheck` renamed and moved to #1016
- `ingress-class-rules` replaced with `certificate-spec-rules`

`certificate-spec-rules` allows to override `issuerRef` and acme solver config based on ingress selector.

**Example**
If you want create certificate for ingress with annotations `{ kubernetes.io/ingress.class: "internal", kubernetes.io/tls-acme: "true" }` and solve acme challenge with ingress class `external` you need to add ConfigMap to cert-manager's namespace with key `certificate-spec-rules` and value:
```json
[{
  "selector": {
    "matchIngressClasses": ["internal"]
  },
  "spec": {
    "issuerRef": {"kind": "ClusterIssuer", "name": "letsencrypt-production"},
    "acme": {
      "http01": {"ingressClass": "external"}
    }
  }
}]
```
`issuerRef` is not required if you don't want to override default issuer.

**Important**
This is not used in production yet, I just want to know your opinion about this concept and to see if there are other people interested in this feature.

munnerz commented:
I don't think we'll accept this PR in its current form, as we don't want to rely on additional external state in this kind of way.

It sounds like there's a few different ways to achieve what you're trying to do, and I'd rather go after those in your other PRs than complicate the API surface with this feature.
